packages/whenable/src/whenable.js

```diff
@@ -0,0 +1,131 @@
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ */
+export const prepareWhenableKits = zone => {
+  /** WeakMap<object, any> */
+  const whenable0ToEphemeral = new WeakMap();
+
+  /**
+   * Get the current incarnation's promise kit associated with a whenable0.
+   *
+   * @param {import('./types.js').WhenablePayload['whenable0']} whenable0
+   * @returns {import('@endo/promise-kit').PromiseKit<any>}
+   */
+  const findCurrentKit = whenable0 => {
+    let pk = whenable0ToEphemeral.get(whenable0);
+    if (pk) {
+      return pk;
+    }
+
+    pk = makePromiseKit();
+    pk.promise.catch(() => {}); // silence unhandled rejection
+    whenable0ToEphemeral.set(whenable0, pk);
+    return pk;
+  };
+
+  /**
+   * @param {'resolve' | 'reject'} kind
+   * @param {import('./types.js').WhenablePayload['whenable0']} whenable0
+   * @param {unknown} value
+   */
+  const settle = (kind, whenable0, value) => {
+    const kit = findCurrentKit(whenable0);
+    const cb = kit[kind];
+    if (!cb) {
+      return;
+    }
+    whenable0ToEphemeral.set(whenable0, harden({ promise: kit.promise }));
+    cb(value);
+  };
+
+  const rawMakeWhenableKit = zone.exoClassKit(
+    'Whenable0Kit',
+    {
+      whenable0: M.interface('Whenable0', {
+        shorten: M.call().returns(M.promise()),
+      }),
+      settler: M.interface('Settler', {
+        resolve: M.call().optional(M.any()).returns(),
+        reject: M.call().optional(M.any()).returns(),
+      }),
+    },
+    () => ({}),
+    {
+      whenable0: {
+        /**
+         * @returns {Promise<any>}
+         */
+        shorten() {
+          return findCurrentKit(this.facets.whenable0).promise;
+        },
+      },
+      settler: {
+        /**
+         * @param {any} [value]
+         */
+        resolve(value) {
+          const { whenable0 } = this.facets;
+          settle('resolve', whenable0, value);
+        },
+        /**
+         * @param {any} [reason]
+         */
+        reject(reason) {
+          const { whenable0 } = this.facets;
+          settle('reject', whenable0, reason);
+        },
+      },
+    },
+  );
+
+  /**
+   * @template T
+   * @returns {import('./types.js').WhenableKit<T>}
+   */
+  const makeWhenableKit = () => {
+    const { settler, whenable0 } = rawMakeWhenableKit();
+    const whenable = makeTagged('Whenable', harden({ whenable0 }));
+    return harden({ settler, whenable });
+  };
+
+  /**
+   * @template T
+   * @returns {import('./types.js').WhenablePromiseKit<T>}
+   */
+  const makeWhenablePromiseKit = () => {
+    const { settler, whenable0 } = rawMakeWhenableKit();
+    const whenable = makeTagged('Whenable', harden({ whenable0 }));
+
+    /**
+     * It would be nice to fully type this, but TypeScript gives:
+     * TS1320: Type of 'await' operand must either be a valid promise or must not contain a callable 'then' member.
+     * @type {unknown}
+     */
+    const whenablePromiseLike = {
+      then(onFulfilled, onRejected) {
```

Comment:
Surprised to see a thenable in the picture. I was hoping to just have whenables and genuine promises, and dispense with thenables completely. Are these ever exposed to users of the API? It looks like they are. Can we avoid this? Should we?

Reply:
This api allows a simple migration to consume whenables from:

```js
const result = await E(obj)[method](...args);
```

to

```js
const result = await when(E(obj)[method](...args));
```

If `when` doesn't supply a Promise API, code will break.  The migration from `when` to `watch` to allow the consumer to be upgradable is an even bigger commitment, since it requires eliminating all `await`s.

Reply:
In that position, the whenablePromiseLike immediately gets absorbed by the `await` to internally create a promise. IOW, in this scenario, the `then` method is immediate invoked, to use the callback behavior to resolve (hopefully settle) the awaited promise. So what is gained by using a promise-like, rather than just making an actual promise that's behaviorally equivalent to `Promise.resolve(whenablePromiseLike)` ?

The only think I can think of is other scenarios where it is worth postponing the absorption behavior. Are such scenarios such a significant concern it is worth introducing thenables? Could a handled promise somehow accommodate the postponement you have in mind?

Reply:
I agree that the postponement is overcomplicated.  It was motivated by a time when I always returned the promise as part of a `makeWhenableKit`, so making it lazy was a useful optimization to avoid keeping a bunch of useless promises around.

Now that `makeWhenablePromiseKit` is separate, I will use your simpler implementation.

Reply:
Sounds like its fixed. Resolving.